examples/game/02-woodeneye-008/woodeneye-008.c

```diff
@@ -0,0 +1,489 @@
+            if (index >= 0) {
+                players[index].yaw -= ((int)event->motion.xrel) * 0x00080000;
+                players[index].pitch = SDL_max(-0x40000000, SDL_min(0x40000000, players[index].pitch - ((int)event->motion.yrel) * 0x00080000));
+            } else if (id) {
+                for (i = 0; i < MAX_PLAYER_COUNT; i++) {
+                    if (players[i].mouse == 0) {
+                        players[i].mouse = id;
+                        as->player_count = SDL_max(as->player_count, i + 1);
+                        break;
+                    }
+                }
+            }
+            break;
+        }
+        case SDL_EVENT_MOUSE_BUTTON_DOWN: {
+            SDL_MouseID id = event->button.which;
+            int index = whoseMouse(id, players, player_count);
+            if (index >= 0) {
+                shoot(index, players, player_count);
+            }
+            break;
+        }
+        case SDL_EVENT_KEY_DOWN: {
+            SDL_Keycode sym = event->key.key;
+            SDL_KeyboardID id = event->key.which;
+            int index = whoseKeyboard(id, players, player_count);
+            if (index >= 0) {
+                if (sym == SDLK_W) players[index].wasd |= 1;
+                if (sym == SDLK_A) players[index].wasd |= 2;
+                if (sym == SDLK_S) players[index].wasd |= 4;
+                if (sym == SDLK_D) players[index].wasd |= 8;
+                if (sym == SDLK_SPACE) players[index].wasd |= 16;
+            } else if (id) {
+                for (i = 0; i < MAX_PLAYER_COUNT; i++) {
+                    if (players[i].keyboard == 0) {
+                        players[i].keyboard = id;
+                        as->player_count = SDL_max(as->player_count, i + 1);
+                        break;
+                    }
+                }
+            }
+            break;
+        }
+        case SDL_EVENT_KEY_UP: {
+            SDL_Keycode sym = event->key.key;
+            SDL_KeyboardID id = event->key.which;
+            if (sym == SDLK_ESCAPE) return SDL_APP_SUCCESS;
+            int index = whoseKeyboard(id, players, player_count);
+            if (index >= 0) {
+                if (sym == SDLK_W) players[index].wasd &= 30;
+                if (sym == SDLK_A) players[index].wasd &= 29;
+                if (sym == SDLK_S) players[index].wasd &= 27;
+                if (sym == SDLK_D) players[index].wasd &= 23;
+                if (sym == SDLK_SPACE) players[index].wasd &= 15;
+            }
+            break;
+        }
+    }
+    return SDL_APP_CONTINUE;
+}
+
+SDL_AppResult SDL_AppIterate(void *appstate)
+{
+    static Uint64 accu = 0;
+    static Uint64 last = 0;
+    static Uint64 past = 0;
+    AppState *as = appstate;
+    Uint64 now = SDL_GetTicksNS();
+    Uint64 dt_ns = now - past;
+    update(as->players, as->player_count, dt_ns);
+    draw(as->renderer, (const float (*)[6])as->edges, as->players, as->player_count);
+    if (now - last > 999999999) {
```

Comment:
Are the fps accurate? Where do the magic values ``999999999`` and ``999999`` come from? What do they mean?

Reply:
Thanks for the explanation. So ``999999999`` means one second in this context. So the fps counter it self is at 1 fps.

And the ``999999`` is used as a fps cap right? If the fps are higher than 1000 (``999999999 / 999999 = 1000``) we stop updating and sleep to not waste resources.

Seems like a useful snippet of code for many sdl projects that is non trivial to understand. Imo this could use a short comment explaining what it is doing. Like ``// limit fps at 1000`` and ``// update fps once a second``